pkg/logger/gorm.go

```diff
@@ -1,95 +1,79 @@
 	logger Logger
 }
 
-// Print formats & prints the log
-func (gl gormLogger) Print(values ...interface{}) {
-	if values[0] == "sql" {
-		location := values[1].(string)
-		duration := values[2].(time.Duration)
-		sql := values[3].(string)
-		formattedValues := formatValues(values[4].([]interface{}))
-		affectedRows := values[5].(int64)
-
-		logger := gl.logger.WithFields(Fields{
-			"sql": Fields{
-				"duration":      duration,
-				"affected_rows": affectedRows,
-				"file_location": location,
-				"values":        formattedValues,
-			},
-		})
-
-		logger.Debugf(sql)
-	} else {
-		gl.logger.Tracef("%v", values[2:]...)
-	}
+func (g gormLogger) LogMode(l logger.LogLevel) logger.Interface {
+	g.logger.SetLevel(convertToLoggerLevel(l))
+
+	return g
+}
+
+func (g gormLogger) Info(ctx context.Context, msg string, args ...interface{}) {
+	g.logger.Infof(msg, args...)
 }
 
-// Code blatantly stolen and modified for our needs from Gorm:
-// https://github.com/jinzhu/gorm/blob/2a3ab99/logger.go#L30-L109
-func formatValues(values []interface{}) string {
-	var formattedValues []string
-	for _, value := range values {
-		indirectValue := reflect.Indirect(reflect.ValueOf(value))
-		if indirectValue.IsValid() {
-			value = indirectValue.Interface()
-			if t, ok := value.(time.Time); ok {
-				if t.IsZero() {
-					formattedValues = append(formattedValues, fmt.Sprintf("'%v'", "0000-00-00 00:00:00"))
-				} else {
-					formattedValues = append(formattedValues, fmt.Sprintf("'%v'", t.Format("2006-01-02 15:04:05")))
-				}
-			} else if b, ok := value.([]byte); ok {
-				if str := string(b); isPrintable(str) {
-					formattedValues = append(formattedValues, fmt.Sprintf("'%v'", str))
-				} else {
-					formattedValues = append(formattedValues, "'<binary>'")
-				}
-			} else if r, ok := value.(driver.Valuer); ok {
-				if value, err := r.Value(); err == nil && value != nil {
-					formattedValues = append(formattedValues, fmt.Sprintf("'%v'", value))
-				} else {
-					formattedValues = append(formattedValues, "NULL")
-				}
-			} else {
-				switch value.(type) {
-				case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, float32, float64, bool:
-					formattedValues = append(formattedValues, fmt.Sprintf("%v", value))
-				default:
-					formattedValues = append(formattedValues, fmt.Sprintf("'%v'", value))
-				}
-			}
-		} else {
-			formattedValues = append(formattedValues, "NULL")
-		}
+func (g gormLogger) Warn(ctx context.Context, msg string, args ...interface{}) {
+	g.logger.Warnf(msg, args...)
+}
+
+func (g gormLogger) Error(ctx context.Context, msg string, args ...interface{}) {
+	g.logger.Errorf(msg, args...)
+}
+
+func (g gormLogger) Trace(ctx context.Context, begin time.Time, fc func() (string, int64), err error) {
+	sql, rows := fc()
+
+	tracingLogger := g.logger.WithFields(Fields{
+		GormTraceFieldKey: Fields{
+			"elapsed seconds": time.Since(begin).Seconds(),
+			"affected rows":   rows,
+			"sql":             sql,
```

Comment:
We are using BindVar everywhere and they must show up as `?`, I did try it locally and can approve that, unless we are passing it in raw query. If this is the case, please update me so I can get it fix ASAP to avoid possible sql injections.

Reply:
when you do a `Exec("select * from test where id = ?", 1)` your'e using bindvar. same thing applies to preparing statements.

Reply:
@Neurostep / @laynax 

If I get the code right this is the `Trace` log level right. SQL would show up only for this level, right?

@Neurostep if SDK consumers are [adding (formatting) user-provided values directly into the query](https://gorm.io/docs/security.html), bypassing security (best) practices, it's on them. The same applies to users of ActiveRecord in the `rails-chassis`.

Reply:
@fotos This is where things are becoming blurry 😅  The logger's `Trace` function is called from 2 places: [db.Scan](https://github.com/go-gorm/gorm/blob/e8f48b5c155b6fbf2e1fe6a554e2280f62af21a7/finisher_api.go#L509) and [processor.Execute](https://github.com/go-gorm/gorm/blob/9fd73ae4f1f638e4c49ae4e6fab8beb9863adabc/callbacks.go#L134) meaning it will be called many times (?). The design of the `Trace` function is a bit weird as it includes also `err` argument that we have to check. So, `Trace` call might produce `Error` log entry and `Debug` log entry depending on a log level configured and if we have `err == nil` or not. gorm's default logger [does almost the same](https://github.com/go-gorm/gorm/blob/08aa2f9888dcd3c950943d09d0d7aaef1b1dcc33/logger/logger.go#L152).